grammar.js

```diff
@@ -747,63 +809,72 @@ module.exports = grammar({
 				$.variable_declaration,
 				"="
 			)),
-			$._expression,
+			field("expression", $._expression),
 			")",
 		),
 
 		when_expression: $ => seq(
 			"when",
-			optional($.when_subject),
+			field("subject", optional($.when_subject)),
 			"{",
-			repeat($.when_entry),
+			field("entries", repeat($.when_entry)),
 			"}"
 		),
 
 		when_entry: $ => seq(
```

Comment:
`when` syntax in Kotlin can contain more than one expression as condition for each entry, for example:

```
when (x) {
            0, 1 -> A()  
            2 -> C()
            else -> B()
        }
```

In UAST,  each `switch cases`   can only contain one expression as condition:

```
class Switch(ConditionalStatement):
    subject: Expression
    cases: Pairs[Expression, Block]
    default: Block
```

So I use a `split_conditions` to convert tree-sitter tree into UAST:

```
    def when_entry(self):
        def split_conditions(node, env):
            conditions = []
            if isinstance(node.conditions, list):
                for condition in node.conditions:
                    conditions.append([condition, node.body])
            else:
                conditions = [[node.conditions, node.body]]
            return conditions

        return choice(
            _(conditions=_) >> build(lambda *x: split_conditions(x[0], x[1])),
            _
        )
```

Because it's a bit complicated, so maybe use a specific `when_entry` can make the rewriter code more readable :D

Reply:
We need to model deconstruction in the AST, we have the same problem with assignments with deconstruction in various languages: `a, b = (1, 2)`